python/ray/data/impl/arrow_serialization.py

```diff
@@ -0,0 +1,437 @@
+        assert len(children) == 1
+        new_children = []
+        for child in children:
+            child = child.slice(child_offset, child_length)
+            new_children.append(_copy_array_if_needed(child))
+        new_buffers.append(buf)
+    elif pa.types.is_struct(type_) or pa.types.is_union(type_):
+        # Dedicated path for StructArrays and UnionArrays.
+        # StructArrays have a top-level bitmap buffer and one or more children arrays.
+        # UnionArrays have a top-level bitmap buffer and type code buffer, and one or
+        # more children arrays.
+        assert children is not None
+        assert buffers is None
+        if pa.types.is_union(type_):
+            # Only sparse unions are supported.
+            assert type_.mode == "sparse"
+            assert buf is not None
+            buf = _copy_buffer_if_needed(buf, pa.int8(), offset, length)
+            new_buffers.append(buf)
+        else:
+            assert buf is None
+        # We copy the children arrays.
+        new_children = []
+        for child in children:
+            new_children.append(_copy_array_if_needed(child))
+    elif pa.types.is_string(type_) or pa.types.is_large_string(type_):
+        # Dedicated path for StringArrays.
+        assert len(buffers) == 1
+        # StringArray buffer scheme: [bitmap, value_offsets, data]
+        offset_buf, data_offset, data_length = _copy_offsets_buffer_if_needed(
+            buf, type_, offset, length
+        )
+        data_buf = _copy_buffer_if_needed(buffers[0], None, data_offset, data_length)
+        new_buffers.append(offset_buf)
+        new_buffers.append(data_buf)
+    else:
+        # If not a nested Array, buf is a plain data buffer.
+        # Copy data buffer, if needed.
+        if buf is not None:
+            buf = _copy_buffer_if_needed(buf, type_, offset, length)
+        new_buffers.append(buf)
+    return pa.Array.from_buffers(type_, length, new_buffers, children=new_children)
+
+
+def _copy_buffer_if_needed(
+    buf: "pyarrow.Buffer",
+    type_: Optional["pyarrow.DataType"],
+    offset: int,
+    length: int,
+) -> "pyarrow.Buffer":
+    """Copy buffer, if needed."""
+    import pyarrow as pa
+
+    if type_ is not None and pa.types.is_boolean(type_):
+        # Arrow boolean array buffers are bit-packed, with 8 entries per byte,
+        # and are accessed via bit offsets.
+        buf = _copy_bitpacked_buffer_if_needed(buf, offset, length)
+    else:
+        type_bytewidth = type_.bit_width // 8 if type_ is not None else 1
+        buf = _copy_normal_buffer_if_needed(buf, type_bytewidth, offset, length)
+    return buf
+
+
+def _copy_normal_buffer_if_needed(
+    buf: "pyarrow.Buffer",
+    byte_width: int,
+    offset: int,
+    length: int,
+) -> "pyarrow.Buffer":
+    """Copy buffer, if needed."""
+    byte_offset = offset * byte_width
+    byte_length = length * byte_width
+    if offset > 0 or byte_length < buf.size:
```

Comment:
Actually no, since we should always have that `byte_offset + byte_length == buf.size`. `offset > 0` is covering the non-zero offset case, and `byte_length < buf.size` is covering the truncated view case (from 0-offset).